File: oc/CustomCell/CustomCell/view/MicroBlogCell.m
//
//  MicroBlogCell.m
//  CustomCell
//
//  Created by dfang on 2017-12-13.
//  Copyright © 2017年 com.east.customcell. All rights reserved.
//

#import "MicroBlogCell.h"

#include "Constants.h"

@interface MicroBlogCell () {

}

@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *descLabel;
@property (nonatomic, strong) UIImageView *iconImageView;
@property (nonatomic, strong) UIImageView *pictureImageView;
@property (nonatomic, strong) UIImageView *vipImageView;

@end

@implementation MicroBlogCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

+ (instancetype)microBlogCellWithTableView:(UITableView *)tableView andCellStyle:(UITableViewCellStyle)cellStyle{
    MicroBlogCell *cell = [tableView dequeueReusableCellWithIdentifier: microBlogCellReuseId];
    if (!cell) {
        cell = [[MicroBlogCell alloc] initWithStyle:cellStyle reuseIdentifier:microBlogCellReuseId];
    }
    return cell;
}

// override
- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:microBlogCellReuseId];
    if (self) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont systemFontOfSize:MicroBlogTitleFont];
        [self.contentView addSubview:_titleLabel];
        
        _descLabel = [[UILabel alloc] init];
        _descLabel.numberOfLines = 0; // 自动换行
        _descLabel.font = [UIFont systemFontOfSize:MicroBlogDescFont];
        [self.contentView addSubview:_descLabel];
        
        _iconImageView = [[UIImageView alloc] init];
        [self.contentView addSubview:_iconImageView];
        
        _pictureImageView = [[UIImageView alloc] init];
        [self.contentView addSubview:_pictureImageView];
        
        _vipImageView = [[UIImageView alloc] init];
        _vipImageView.image = [UIImage imageNamed:@"vip"];
        [self.contentView addSubview:_vipImageView];
    }
    return self;
}

- (void)setMicroBlog:(MicroBlog *)microBlog {
    _microBlog = microBlog;
    
    // set subviews content
    _titleLabel.text = _microBlog.name;
    _descLabel.text = _microBlog.text;
    _iconImageView.image = [UIImage imageNamed:_microBlog.icon];
    if (_microBlog.picture) {
        _pictureImageView.image = [UIImage imageNamed:_microBlog.picture];
    }
    
    if (_microBlog.isVip) {
        _vipImageView.hidden = NO;
        _titleLabel.textColor = [UIColor redColor];
    } else {
        _vipImageView.hidden = YES;
        _titleLabel.textColor = [UIColor blackColor];
    }
    
    // set subviews frame
    _titleLabel.frame = _microBlog.nameFrame;
    _descLabel.frame = _microBlog.textFrame;
    _iconImageView.frame = _microBlog.iconFrame;
    _pictureImageView.frame = _microBlog.pictureFrame;
    _vipImageView.frame = _microBlog.vipFrame;
}

@end

File: swift/CustomCell/CustomCell/view/MicroBlogCell.swift
//
//  MicroBlogCell.swift
//  CustomCell
//
//  Created by dfang on 2017-12-14.
//  Copyright © 2017年 com.east.customcell. All rights reserved.
//

import Foundation
import UIKit

class MicroBlogCell: UITableViewCell{
    lazy var nameLabel: UILabel = UILabel.init()
    lazy var descLabel: UILabel = UILabel.init()
    lazy var iconImageView: UIImageView = UIImageView.init()
    lazy var pictureImageView: UIImageView = UIImageView.init()
    lazy var vipImageView: UIImageView = UIImageView.init()
    var _microBlogFrame: MicroBlogFrame = MicroBlogFrame.init()
    
    var microBlogFrame: MicroBlogFrame {
        get {
            return _microBlogFrame
        }
        set (newValue){
            _microBlogFrame = newValue
            _microBlogFrame.setMicroBlog(_microBlog: newValue.microBlog)
            
            // set subviews content
            let microBlog:MicroBlog = _microBlogFrame.microBlog
            nameLabel.text = microBlog.name
            descLabel.text = microBlog.text
            iconImageView.image = UIImage.init(named: microBlog.icon!)
            if (nil != microBlog.picture) {
                pictureImageView.image = UIImage.init(named: microBlog.picture!)
            }
            if microBlog.vip {
                vipImageView.isHidden = false
                nameLabel.textColor = UIColor.red
            } else {
                vipImageView.isHidden = true
                nameLabel.textColor = UIColor.black
            }
            
            // set subviews frame
            nameLabel.frame = _microBlogFrame.nameFrame
            descLabel.frame = _microBlogFrame.textFrame
            iconImageView.frame = _microBlogFrame.iconFrame
            pictureImageView.frame = _microBlogFrame.pictureFrame
            vipImageView.frame = _microBlogFrame.vipFrame
        }
    }
    
    class func microBlogCellWithTableView(tableView: UITableView) -> MicroBlogCell {
        var cell: MicroBlogCell? = tableView.dequeueReusableCell(withIdentifier: MICROBLOGCELLREUSEID) as? MicroBlogCell
        if (nil == cell) {
            cell = MicroBlogCell.init(style: UITableViewCellStyle.default, reuseIdentifier: MICROBLOGCELLREUSEID)
        }
        
        return cell!
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.addContent()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func addContent() {
        nameLabel.font = UIFont.systemFont(ofSize: MICROBLOGNAMEFONTSIZE)
        self.contentView.addSubview(nameLabel)
        
        descLabel.font = UIFont.systemFont(ofSize: MICROBLOGTEXTFONTSIZE)
        descLabel.numberOfLines = 0
        self.contentView.addSubview(descLabel)
    
        self.contentView.addSubview(iconImageView)
        self.contentView.addSubview(pictureImageView)
        
        vipImageView.image = UIImage.init(named: "vip")
        self.contentView.addSubview(vipImageView)
    }
}
